src/pool/cli.c

```diff
@@ -184,36 +184,53 @@ dc_pool_alloc(unsigned int nr)
 	return NULL;
 }
 
-/* rsvc_client callback function to ask mgmt. service for updated
- * list of pool service replica ranks
+/* Choose a pool service replica rank. If the rsvc module indicates
+ * DER_NOTREPLICA, attempt to refresh the list by querying the MS.
  */
-static int
-pool_rsvc_update_ranks_cb(struct rsvc_client *client)
+int
+dc_pool_choose_svc_rank(const uuid_t puuid, struct rsvc_client *cli,
+			pthread_mutex_t *cli_lock, struct dc_mgmt_sys *sys,
+			crt_endpoint_t *ep)
 {
-	struct dc_mgmt_sys     *sys = (struct dc_mgmt_sys *)client->sc_priv;
-	d_rank_list_t	       *new_ranklist = NULL;
-	int			i;
 	int			rc;
+	int			i;
 
-	D_DEBUG(DF_DSMC, DF_UUID": request updated replica rank list from MS\n",
-		DP_UUID(client->sc_id));
-
-	rc = dc_mgmt_get_pool_svc_ranks(sys, client->sc_id, &new_ranklist);
-	if (rc != 0) {
-		D_ERROR(DF_UUID ": dc_mgmt_get_pool_svc_ranks() failed, "
-			DF_RC "\n", DP_UUID(client->sc_id), DP_RC(rc));
-		return rc;
-	}
-
-	rc = d_rank_list_copy(client->sc_ranks, new_ranklist);
-	if (rc == 0) {
-		for (i = 0; i < client->sc_ranks->rl_nr; i++) {
-			D_DEBUG(DF_DSMC, DF_UUID ": sc_ranks[%d]=%u\n",
-				DP_UUID(client->sc_id), i,
-				client->sc_ranks->rl_ranks[i]);
+	if (cli_lock)
+		D_MUTEX_LOCK(cli_lock);
```

Comment:
Out of curiosity, when could this lock be NULL? Is it really OK to proceed with the rest of the function body if there's no lock?

Reply:
Here are the places in the client code that use their own state struct, and not the struct dc_pool that has its own pthread_mutex_t client lock. I think this happens when the client doesn't do a pool connect.

dc_pool_evict(), 
dc_pool_update_internal() [that is used in pool drain, exclude, exclude_out, reint], 
rsvc_client_state_create() [used by dc_pool_membership_update() used by dc_pool_add/remove_replicas() that are likely deprecated APIs to be removed, and replaced with functionality in the control plane]